### PR TITLE
Update RazorPages, MVC, WebApi templates to use Identity.Web

### LINKED ISSUE #24024: Incorporate Microsoft.Identity.Web templates into our project templates
We want to use the templates at https://github.com/AzureAD/microsoft-identity-web/tree/master/ProjectTemplates/templates to replace our organization auth logic.

### PR DESCRIPTION
**Pull request by JunTaoLuo**
Fixes #24024 
Fixes https://github.com/dotnet/aspnetcore/issues/22726

ported from https://github.com/dotnet/aspnetcore/pull/24065

fyi @jmprieur

### CODE REVIEW

File: src/ProjectTemplates/Web.ProjectTemplates/WebApi-CSharp.csproj.in [hunk 1]
```diff
@@ -3,16 +3,14 @@
...
     <NoDefaultLaunchSettingsFile Condition="'$(ExcludeLaunchSettings)' == 'True'">True</NoDefaultLaunchSettingsFile>
     <RootNamespace Condition="'$(name)' != '$(name{-VALUE-FORMS-}safe_namespace)'">Company.WebApplication1</RootNamespace>
   </PropertyGroup>
-
   <!--#if (OrganizationalAuth || IndividualB2CAuth) -->
+
   <ItemGroup>
-    <PackageReference Include="Microsoft.AspNetCore.Authentication.AzureAD.UI" Version="${MicrosoftAspNetCoreAuthenticationAzureADUIPackageVersion}" Condition="'$(OrganizationalAuth)' == 'True'" />
-    <PackageReference Include="Microsoft.AspNetCore.Authentication.AzureADB2C.UI" Version="${MicrosoftAspNetCoreAuthenticationAzureADB2CUIPackageVersion}" Condition="'$(IndividualB2CAuth)' == 'True'" />
+    <PackageReference Include="Microsoft.Identity.Web" Version="${MicrosoftIdentityWebPackageVersion}" />
```
Comment (Tratcher) [inline src/ProjectTemplates/Web.ProjectTemplates/WebApi-CSharp.csproj.in:12]: Have we worked out what these versions will be like at RTM? This is where the circular dependency starts to worry me.

File: src/ProjectTemplates/Web.ProjectTemplates/content/RazorPagesWeb-CSharp/Services/DownstreamWebApi.cs [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        }
+
+        /// <summary>
+        /// Calls the Web API with the required scopes
+        /// </summary>
+        /// <param name="requireScopes">[Optional] Scopes required to call the Web API. If
+        /// not specified, uses scopes from the configuration</param>
+        /// <param name="relativeEndpoint">Endpoint relative to the CalledApiUrl configuration</param>
+        /// <returns>A JSON string representing the result of calling the Web API</returns>
+        public async Task<string> CallWebApi(string relativeEndpoint = "", string[] requiredScopes = null)
```
Comment (Tratcher) [inline src/ProjectTemplates/Web.ProjectTemplates/content/RazorPagesWeb-CSharp/Services/DownstreamWebApi.cs:49]:
```suggestion
        public async Task<string> CallWebApiAsync(string relativeEndpoint = "", string[] requiredScopes = null)
```

File: src/ProjectTemplates/Web.ProjectTemplates/content/RazorPagesWeb-CSharp/Startup.cs [hunk 3]
```diff
@@ -64,57 +71,46 @@ public void ConfigureServices(IServiceCollection services)
...
 #if (OrganizationalAuth)
 
-            services.AddRazorPages();
             services.AddAuthorization(options =>
             {
                 // By default, all incoming requests will be authorized according to the default policy
                 options.FallbackPolicy = options.DefaultPolicy;
             });
+            services.AddRazorPages()
+                    .AddMvcOptions(options => {})
```
Comment (Tratcher) [inline src/ProjectTemplates/Web.ProjectTemplates/content/RazorPagesWeb-CSharp/Startup.cs:109]: ?

**Review by Tratcher — COMMENTED** [review]
I filed https://github.com/AzureAD/microsoft-identity-web/issues/346 for an AzureAd migration guide, but we'll also need a docs issue for the AspNetCore migration guide that links to theirs.

File: src/ProjectTemplates/Web.ProjectTemplates/content/RazorPagesWeb-CSharp/Services/MicrosoftGraphServiceExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+        /// <param name="graphBaseUrl">Base URL for Microsoft graph. This can be
+        /// changed for instance for applications running in national clouds</param>
+        public static IServiceCollection AddMicrosoftGraph(this IServiceCollection services,
+                                             IEnumerable<string> initialScopes,
+                                             string graphBaseUrl = "https://graph.microsoft.com/v1.0")
+        {
+            services.AddTokenAcquisition(true);
+            services.AddSingleton<GraphServiceClient, GraphServiceClient>(serviceProvider =>
+            {
+                var tokenAquisitionService = serviceProvider.GetService<ITokenAcquisition>();
```
Comment (JunTaoLuo) [inline src/ProjectTemplates/Web.ProjectTemplates/content/RazorPagesWeb-CSharp/Services/MicrosoftGraphServiceExtensions.cs:27]: https://github.com/dotnet/aspnetcore/pull/24065#discussion_r458053335 This probably needs to be GetRequiredService if you expect this to not be null.

File: src/ProjectTemplates/Web.ProjectTemplates/content/RazorPagesWeb-CSharp/Services/DownstreamWebApi.cs [hunk 5]
```diff
@@ -0,0 +1,72 @@
+﻿using System.Net;
+using System.Net.Http;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Identity.Web;
+
+namespace Company.WebApplication1
+{
+    public interface IDownstreamWebApi
```
Comment (JunTaoLuo) [inline src/ProjectTemplates/Web.ProjectTemplates/content/RazorPagesWeb-CSharp/Services/DownstreamWebApi.cs:10]: https://github.com/dotnet/aspnetcore/pull/24065#discussion_r458052024 Is there a specific reason why we need an interface here?

File: src/ProjectTemplates/Web.ProjectTemplates/content/StarterWeb-CSharp/Views/Home/Index.cshtml [hunk 6]
```diff
@@ -6,3 +6,8 @@
     <h1 class="display-4">Welcome</h1>
     <p>Learn about <a href="https://docs.microsoft.com/aspnet/core">building Web apps with ASP.NET Core</a>.</p>
 </div>
+    @* #if(GenerateApiOrGraph)
```
Comment (JunTaoLuo) [inline src/ProjectTemplates/Web.ProjectTemplates/content/StarterWeb-CSharp/Views/Home/Index.cshtml:9]: FYI this doesn't work. The generated templates just has this section verbatim.

Comment (JunTaoLuo) [reply src/ProjectTemplates/Web.ProjectTemplates/content/StarterWeb-CSharp/Views/Home/Index.cshtml]: I tried fixing this with https://github.com/dotnet/aspnetcore/pull/24167/commits/3b093a88569e418fdc3c45340cbc0e26faa46085. It's working as intended but doesn't seem like the right syntax.

File: src/ProjectTemplates/Web.ProjectTemplates/content/StarterWeb-CSharp/Controllers/HomeController.cs [hunk 7]
```diff
@@ -41,10 +41,7 @@ public class HomeController : Controller
         [AuthorizeForScopes(ScopeKeySection = "CalledApi:CalledApiScopes")]
         public async Task<IActionResult> Index()
         {
-            ViewData["ApiResult"] = await _downstreamWebApi.CallWebApi();
```
Comment (JunTaoLuo) [inline src/ProjectTemplates/Web.ProjectTemplates/content/StarterWeb-CSharp/Controllers/HomeController.cs]: Hey I think you missed this @jmprieur

**Comment by JunTaoLuo** [comment]
@jmprieur and @jennyf19 were able to verify that {mvc, webapp, webapi} x {IndividualB2C + API, SingleOrg + Api, SingleOrg + Graph} all works so this is ready to be merged.

File: src/ProjectTemplates/Web.ProjectTemplates/content/RazorPagesWeb-CSharp/Services/DownstreamWebApi.cs [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+
+            string apiResult;
+            var response = await _httpClient.SendAsync(httpRequestMessage);
+            if (response.StatusCode == HttpStatusCode.OK)
+            {
+                apiResult = await response.Content.ReadAsStringAsync();
+            }
+            else
+            {
+                apiResult = $"Error calling the API '{apiUrl}'";
```
Comment (Tratcher) [inline src/ProjectTemplates/Web.ProjectTemplates/content/RazorPagesWeb-CSharp/Services/DownstreamWebApi.cs:66]: Later: Add logging

File: src/ProjectTemplates/Web.ProjectTemplates/content/RazorPagesWeb-CSharp/Startup.cs [hunk 9]
```diff
@@ -64,57 +71,46 @@ public void ConfigureServices(IServiceCollection services)
...
-                    //OnTokenValidated = context =>
-                    //{
-                    //    return myUserValidationLogic(context.Ticket.Principal);
-                    //}
-                };
-            });
+            services.AddMicrosoftWebAppAuthentication(Configuration, "AzureAd")
+#if (GenerateApiOrGraph)
+                    .AddMicrosoftWebAppCallsWebApi(Configuration,
+                                                   "AzureAd")
```
Comment (Tratcher) [inline src/ProjectTemplates/Web.ProjectTemplates/content/RazorPagesWeb-CSharp/Startup.cs:77]:
```suggestion
                    .AddMicrosoftWebAppCallsWebApi(Configuration, "AzureAd")
```

**Review by Tratcher — APPROVED** [review]

File: src/ProjectTemplates/Web.ProjectTemplates/content/RazorPagesWeb-CSharp/Startup.cs (hunk 9 above)
Comment (JunTaoLuo) [reply src/ProjectTemplates/Web.ProjectTemplates/content/RazorPagesWeb-CSharp/Startup.cs:77]: I'll address this in the Blazor templates PR, this appears in several places IIRC.

**Comment by satishrabari** [comment]
Hi, Would you please help me with the following things?
I am using Microsoft Identity Platform.
I want to redirect to account/logincallback URL after successful login.
Where I put the redirect URL for this?

I want to configure using startup.cs because I want to save user info in DB and add custom claims in the claim principle.
Is it possible?
I don't have a login URL on my web.

following is my Startup.cs
` 

            services.AddAuthentication(OpenIdConnectDefaults.AuthenticationScheme)
                .AddMicrosoftIdentityWebApp(Configuration.GetSection("AzureAd"));

            services.AddControllersWithViews(options =>
            {
                var policy = new AuthorizationPolicyBuilder()
                    .RequireAuthenticatedUser()
                    .Build();
                options.Filters.Add(new AuthorizeFilter(policy));
            });
            services.AddRazorPages()
                 .AddMicrosoftIdentityUI();
       

`

**Pull request merged**